### PR TITLE
#325 Linter rechecks files if they were listed more than once

### LINKED ISSUE #325: Linter rechecks files if they were listed more than once
Create 3 files in one dir:
```
$ mkdir -p /tmp/project
$ for f in a b c; do echo "ohno" > /tmp/project/.env.$f; done
$ tree -a /tmp/project
/tmp/project
├── .env.a
├── .env.b
└── .env.c
```

Listing whole dir twice leads to each file being checked twice:
```
$ cargo run -- /tmp/project/ /tmp/project/
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
     Running `target/debug/dotenv-linter /tmp/project/ /tmp/project/`
../../../../tmp/project/.env.a:1 KeyWithoutValue: The ohno key should be with a value or have an equal sign
../../../../tmp/project/.env.b:1 KeyWithoutValue: The ohno key should be with a value or have an equal sign
../../../../tmp/project/.env.c:1 KeyWithoutValue: The ohno key should be with a value or have an equal sign
../../../../tmp/project/.env.a:1 KeyWithoutValue: The ohno key should be with a value or have an equal sign
../../../../tmp/project/.env.b:1 KeyWithoutValue: The ohno key should be with a value or have an equal sign
../../../../tmp/project/.env.c:1 KeyWithoutValue: The ohno key should be with a value or have an equal sign

Found 6 problems
```

Listing file `.env.a` and whole dir works as expected:
```
$ cargo run -- /tmp/project/ /tmp/project/.env.a
    Finished dev [unoptimized + debuginfo] target(s) in 0.03s
     Running `target/debug/dotenv-linter /tmp/project/ /tmp/project/.env.a`
../../../../tmp/project/.env.a:1 KeyWithoutValue: The ohno key should be with a value or have an equal sign
../../../../tmp/project/.env.b:1 KeyWithoutValue: The ohno key should be with a value or have an equal sign
../../../../tmp/project/.env.c:1 KeyWithoutValue: The ohno key should be with a value or have an equal sign

Found 3 problems
```
Hovewer, listing file `.env.b` and whole dir leads to the file being checked twice:
```
$ cargo run -- /tmp/project/ /tmp/project/.env.b
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
     Running `target/debug/dotenv-linter /tmp/project/ /tmp/project/.env.b`
../../../../tmp/project/.env.b:1 KeyWithoutValue: The ohno key should be with a value or have an equal sign
../../../../tmp/project/.env.a:1 KeyWithoutValue: The ohno key should be with a value or have an equal sign
../../../../tmp/project/.env.b:1 KeyWithoutValue: The ohno key should be with a value or have an equal sign
../../../../tmp/project/.env.c:1 KeyWithoutValue: The ohno key should be with a value or have an equal sign

Found 4 problems
```

I guess the issue is here https://github.com/dotenv-linter/dotenv-linter/blob/281a2030aa66bd3d3dcbbabb1636659548e96a98/src/lib.rs#L101-L103
Sorted file names extended with file names from dirs and then *consecutive* duplicates removed.

In case with file `.env.a`:
```
# file names
/tmp/project/.env.a
# file names from dirs
/tmp/project/.env.a <- removed by deduplication
/tmp/project/.env.b
/tmp/project/.env.c
```

In case with file `.env.b`:
```
# file names
/tmp/project/.env.b
# file names from dirs
/tmp/project/.env.a
/tmp/project/.env.b <- isn't removed because there are no consecutive duplicates here
/tmp/project/.env.c
```

Switching order of sorting and extending should fix the issue. Yay, 100 line long babbling for 2 line change in code!

### PR DESCRIPTION
**Pull request by Aashu23**
@mgrachev 
Fixed the issue
Updated tests accordingly

Fixes #325 

#### ✔ Checklist:

- [x] This PR has been added to [CHANGELOG.md](https://github.com/dotenv-linter/dotenv-linter/blob/master/CHANGELOG.md) (at the top of the list);
- [x] Tests for the changes have been added (for bug fixes / features);

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Add action-yamllint [#317](https://github.com/dotenv-linter/dotenv-linter/pull/317) ([@vk26](https://github.com/vk26))
 
 ### 🔧 Changed
+- Fix Linter rechecking files if they were listed more than once [#327](https://github.com/dotenv-linter/dotenv-linter/pull/327) ([@aashu23](https://github.com/Aashu23))
```
Comment (mgrachev) [inline CHANGELOG.md:12]:
```suggestion
- Fix linter rechecking files if they were listed more than once [#327](https://github.com/dotenv-linter/dotenv-linter/pull/327) ([@aashu23](https://github.com/Aashu23))
```

**Review by mgrachev — COMMENTED** [review]
@Aashu23 👋 Thank you for your contribution! 🚀

I have left a comment. Please take a look 👀
Also, please add two intergration tests to check that all works as expected:
1. Case 1 `$ dotenv-linter /path1 /path1`;
2. Case 2 `$ dotenv-linter /path1 /path1/.env.b`

File: CHANGELOG.md [hunk 2]
```diff
@@ -9,6 +9,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Add action-yamllint [#317](https://github.com/dotenv-linter/dotenv-linter/pull/317) ([@vk26](https://github.com/vk26))
 
 ### 🔧 Changed
+- Fix linter rechecking files if they were listed more than once [#327](https://github.com/dotenv-linter/dotenv-linter/pull/327) ([@aashu23](https://github.com/Aashu23))
```
Comment (mgrachev) [inline CHANGELOG.md:12]:
```suggestion
- Fix linter rechecking files if they were listed more than once [#327](https://github.com/dotenv-linter/dotenv-linter/pull/327) ([@Aashu23](https://github.com/Aashu23))
```

File: tests/args/specific_path.rs [hunk 3]
```diff
@@ -82,6 +82,52 @@ fn checks_two_specific_files() {
     testdir.test_command_fail_with_args(args, expected_output);
 }
 
+#[test]
+fn checks_each_file_only_once_when_listing_same_path_twice() {
+    let testdir = TestDir::new();
+    testdir.create_testfile(".env", "foo=\n");
```
Comment (mgrachev) [inline tests/args/specific_path.rs:88]:
```suggestion
```
Why this file is needed? You don't use it in the test case.

File: tests/args/specific_path.rs [hunk 4]
```diff
@@ -82,6 +82,52 @@ fn checks_two_specific_files() {
...
+
+    let subdir = testdir.subdir();
+    let testfile_1 = subdir.create_testfile(".env", " FOO=\n");
+    let testfile_2 = subdir.create_testfile(".env", " FOO=val\nBAR=foo\n");
```
Comment (mgrachev) [inline tests/args/specific_path.rs:92]: You can't create files with the same name in the same directory.

File: tests/args/specific_path.rs [hunk 5]
```diff
@@ -82,6 +82,52 @@ fn checks_two_specific_files() {
...
+            .to_str().expect("multi-platform path to test .env file")
+    );
+
+    testdir.test_command_fail_with_args(args, expected_output);
+}
+
+#[test]
+fn checks_each_file_only_once_when_listing_one_path_and_one_file() {
+    let testdir = TestDir::new();
+    testdir.create_testfile(".env", "foo=\n");
```
Comment (mgrachev) [inline tests/args/specific_path.rs:111]:
```suggestion
```
Why this file is needed? You don't use it in the test case.

File: tests/args/specific_path.rs [hunk 6]
```diff
@@ -82,6 +82,52 @@ fn checks_two_specific_files() {
...
+
+    let subdir = testdir.subdir();
+    let testfile_1 = subdir.create_testfile(".env", " FOO=\n");
+    let testfile_2 = subdir.create_testfile(".env", " FOO=val\nBAR=foo\n");
```
Comment (mgrachev) [inline tests/args/specific_path.rs:115]: You can't create files with the same name in the same directory.

File: tests/args/specific_path.rs (hunk 4 above)
Comment (Aashu23) [reply tests/args/specific_path.rs:92]: It automatically created as env2. I can make it explicit though.

**Review by mgrachev — APPROVED** [review]

**Comment by mgrachev** [comment]
@Aashu23 Thank you! 👍

🙏 If it’s not difficult for you, please support the project - click on the star on GitHub ⭐️

**Pull request merged**